File: ctbrpgsca/state.cpp
/*
Copyright (C) AD 2013-2020 Claudiu-Stefan Costea

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
*/
#include "state.h"
#include "ability.h"
#include "costume.h"
#include "actor.h"
#include "scene.h"
#include "role.h"

#include <QStringBuilder>

using namespace tbrpgsca;

const Ability& State::removedSkill(int const n) const
{
    return *(this->_r_skills->at(n));
}

bool State::hasRemovedSkill(const Ability& skill) const
{
    QVector<const Ability*>* aSkills = this->_r_skills;
    return aSkills && aSkills->contains(&skill);
}

int State::removedSkillsSize() const
{
    QVector<const Ability*>* aSkills = this->_r_skills;
    return aSkills ? aSkills->size() : 0;
}

void State::inflict(QString& ret, Actor* user, Actor& target, int dur, const bool always) const
{
    return this->inflict(&ret, nullptr, user, target, dur, always);
}

void State::inflict(QString* const ret, Scene* const scene, Actor* const user, Actor& target, int stateDur, const bool always) const
{
    const State& state = *this;
    if (stateDur == 0)
    {
        stateDur = state._dur;
        if (stateDur == 0)
        {
            return;
        }
    }
    if (stateDur > STATE_END_DUR)
    {
        int stateRes;
        QMap<const State*, int>* stRes;
        if (always || (stateRes = state._s_res) < 0 || ((std::rand() % 10) > (((stRes = target._st_res)
                == nullptr ? 0 : stRes->value(this, 0) + stRes->value(nullptr, 0)) + stateRes)))
        {
            QMap<const State*, int>* trgStates = target._state_dur;
            if (trgStates)
            {
                QMap<const State*, int>* const rStates = state._state_dur;
                if (rStates)
                {
                    auto const rLast = rStates->cend();
                    for (auto rIt = rStates->cbegin(); rIt != rLast; ++rIt)
                    {
                        int rDur = rIt.value();
                        const State* const rState = rIt.key();
                        if (rDur == 0 || rDur <= STATE_END_DUR)
                        {
                            rDur = rState->_dur;
                        }
                        auto const last = trgStates->cend();
                        for (auto it = trgStates->cbegin(); it != last; it += 1)
                        {
                            const State* const aState = it.key();
                            if (aState == rState)
                            {
                                int const aDur = it.value();
                                if (aDur > STATE_END_DUR)
                                {
                                    if (aDur < 0 && rDur > aDur)
                                    {
                                        return;
                                    }
                                    rState->disable(ret, scene, target, rDur, false);
                                    if (rDur > 0 && aDur > 0)
                                    {
                                        stateDur -= aDur < rDur ? aDur : rDur;
                                        if (stateDur < 1)
                                        {
                                            return;
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
            else
            {
                trgStates = new QMap<const State*, int>();
                target._state_dur = trgStates;
            }
            state.blockSkills(target, false);
            int const crDur = trgStates->value(this, STATE_END_DUR);
            if (crDur == STATE_END_DUR)
            {
                state.adopt(ret, scene, target, false, false);
                trgStates->operator[](this) = stateDur;
            }
            else if ((crDur > -1 && crDur < stateDur) || (stateDur < 0 && stateDur < crDur))
            {
                trgStates->operator[](this) = stateDur;
            }
            if (user && this->isConverted() && target._side != user->_side)
            {
                target._side = user->_side;
            }
        }
    }

}

void State::remove(QString* const ret, Scene* const scene, Actor& actor) const
{
    const State& state = *this;
    state.blockSkills(actor, true);
    state.adopt(ret, scene, actor, false, true);
}

bool State::disable(Actor& actor, int const dur, const bool remove) const
{
    return this->disable(nullptr, nullptr, actor, remove, dur);
}

bool State::disable(QString* const ret, Scene* const scene, Actor& actor, int dur, const bool remove) const
{
    if (dur == 0)
    {
        dur = this->_dur;
    }
    if (dur > STATE_END_DUR)
    {
        QMap<const State*, int>* sDur = actor._state_dur;
        if (sDur)
        {
            int const crDur = sDur->value(this, STATE_END_DUR);
            //if (dur == -2 || (crDur > -2 && (dur == -1 || crDur > -1)))
            if (crDur > -1 || dur <= crDur)
            {
                if (dur > 0 /*&& crDur > 0*/ && crDur > dur)
                {
                    sDur->operator[](this) = crDur - dur;
                    return false;
                }
                else
                {
                    if (remove)
                    {
                        sDur->remove(this);
                    }
                    else
                    {
                        sDur->operator[](this) = STATE_END_DUR;
                    }
                    if (crDur > STATE_END_DUR)
                    {
                        this->remove(ret, scene, actor);
                    }
                    return true;
                }
            }
            return crDur <= STATE_END_DUR;
        }
        else
        {
            return true;
        }
    }
    else
    {
        return false;
    }
}

void State::alter(QString& ret, Actor& actor, const bool consume) const
{
    return this->alter(&ret, nullptr, actor, consume);
}

void State::alter(QString* const ret, Scene* const scene, Actor& actor, const bool consume) const
{
    const State& state = *this;
    QMap<const State*, int>* sDur = actor._state_dur;
    if (sDur /*&& actor.hp > 0*/)
    {
        int const d = sDur->value(this, STATE_END_DUR);
        if (consume)
        {
            if (d > 0)
            {
                sDur->operator[](this) = d - 1;
            }
        }
        else if (d == 0)
        {
            sDur->operator[](this) = STATE_END_DUR;
            state.remove(ret, scene, actor);
        }
    }

}

void State::blockSkills(Actor& actor, const bool remove) const
{
    const State& state = *this;
    QVector<const Ability*>* rSkills = state._r_skills;
    if (rSkills)
    {
        QMap<const Ability*, int>* iSkills = actor._skills_cr_qty;
        if (remove)
        {
            if (iSkills)
            {
                for (const Ability* const skill : *rSkills)
                {
                    if (skill->_m_qty > 0)
                    {
                        iSkills->operator[](skill) = -1 * iSkills->value(skill, 0);
                    }
                    else
                    {
                        iSkills->remove(skill);
                    }
                }
            }
        }
        else
        {
            if (iSkills == nullptr)
            {
                iSkills = new QMap<const Ability*, int>();
                actor._skills_cr_qty = iSkills;
            }
            for (const Ability* const skill : *rSkills)
            {
                iSkills->operator[](skill) = skill->_m_qty > 0 ? -1 * iSkills->value(skill, 0) : 0;
            }
        }
    }

}

State::State(int const id, QString name, QString sprite, bool const shapeShift, int const dur, int const sRes, int const mActions, int const elm, int const hpDmg,
             int const mpDmg, int const spDmg, int const mHp, int const mMp, int const mSp, int const atk, int const def, int const spi, int const wis, int const agi,
             bool const stun, bool const range, bool const automate, bool const confuse, bool const convert, bool const reflect, bool const ko, bool const invincible,
             bool const revive, QVector<const Ability*>* const aSkills, QVector<const Ability*>* const counters, QVector<const Ability*>* const rSkills,
             QMap<const State*, int>* const states, QMap<const State*, int>* const stRes, QMap<int, int>* const res)
    : Costume(id, name, sprite, shapeShift, mActions, elm, hpDmg, mpDmg, spDmg, mHp, mMp, mSp, atk, def, spi, wis, agi, stun, range, automate, confuse, reflect, ko,
              invincible, revive, aSkills, counters, states, stRes, res)
{
    if (convert)
    {
        this->_play_flags |= FLAG_CONVERT;
    }
    this->_dur = dur;
    this->_s_res = sRes;
    this->_r_skills = rSkills;
}

State::State(const State& state) : Costume(state)
{
    this->_dur = state._dur;
    this->_s_res = state._s_res;
    this->_r_skills = state._r_skills;
}

State::~State()
{

}

File: itbrpgsca/itbrpgsca/State.swift
/*
Copyright (C) AD 2013-2020 Claudiu-Stefan Costea

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
*/

import Foundation

open class State : Costume {
    
    public static let STATE_END_DUR = -3
    
    internal var _dur: Int, _sRes: Int, _rSkills: [Ability]?
    
    open var dur: Int {
        return self._dur
    }
    
    open var sRes: Int {
        return self._sRes
    }
    
    open var rSkills: [Ability]? {
        return self._rSkills
    }
    
    open func remove(actor: Actor) {
        self.blockSkills(actor: actor, remove: true)
        self.adopt(actor: actor, updStates: false, remove: true)
        if self.converted && !actor.converted {
            actor.side = actor._oldSide
        }
        if self.covered && !actor.covered {
            actor.coveredBy = nil
        }
        if self.drawn && !actor.drawn {
            actor.drawnBy = nil
        }
    }
    
    open func blockSkills(actor: Actor, remove: Bool) {
        if let rSkills = self.rSkills {
            var iSkills: [Ability : Int]! = actor._skillsCrQty
            if remove {
                if iSkills != nil {
                    for skill in rSkills {
                        if skill.mQty > 0 {
                            iSkills[skill] = -1 * (iSkills[skill] ?? 0)
                        } else {
                            iSkills.removeValue(forKey: skill)
                        }
                    }
                }
            } else {
                if iSkills == nil {
                    iSkills = [Ability : Int]()
                    actor._skillsCrQty = iSkills
                }
                for skill in rSkills {
                    iSkills[skill] = skill.mQty > 0 ? -1 * (iSkills[skill] ?? 0) : 0
                }
            }
        }
    }
    
    open func alter(ret: inout String?, actor: Actor, consume: Bool) {
        if var sDur = actor.stateDur {
            let d = sDur[self] ?? State.STATE_END_DUR
            if consume {
                if d > 0 {
                    sDur[self] = d - 1
                }
            } else if dur == 0 {
                sDur[self] = State.STATE_END_DUR
                self.remove(actor: actor)
            }
        }
    }
    
    open func disable(actor: Actor, dur: Int, remove: Bool) -> Bool {
        let d = dur == 0 ? self.dur : dur
        if d > State.STATE_END_DUR {
            if var sDur = actor.stateDur {
                let crDur = sDur[self] ?? State.STATE_END_DUR
                if crDur > -1 || d <= crDur {
                    if d > 0 && crDur > d {
                        sDur[self] = crDur - d
                        return false
                    } else {
                        if remove {
                            sDur.removeValue(forKey: self)
                        } else {
                            sDur[self] = State.STATE_END_DUR
                        }
                        if crDur > State.STATE_END_DUR {
                            self.remove(actor: actor)
                        }
                        return true
                    }
                }
                return crDur <= State.STATE_END_DUR
            } else {
                return true
            }
        } else {
            return false
        }
    }
    
    open func inflict(user: Actor!, target: Actor, dur: Int, always: Bool) {
        var d: Int
        if dur == 0 {
            d = self.dur
            if d == 0 {
                return
            }
        } else {
            d = dur
        }
        if d > State.STATE_END_DUR {
            if !always {
                var stateRes = self.sRes
                if stateRes > -1 {
                    let stRes: [State? : Int]! = target.stRes
                    stateRes = stRes == nil ? 0 : (stRes[self] ?? 0) + (stRes[nil] ?? 0) + stateRes
                    if Int.random(in: 0..<10) < stateRes {
                        return
                    }
                }
            }
            var trgStates: [State : Int]! = target.stateDur
            if trgStates == nil {
                trgStates = [State : Int]()
                target.stateDur = trgStates
            } else if let rStates = self.stateDur {
                for (rState, var rDur) in rStates {
                    if rDur == 0 || rDur <= State.STATE_END_DUR {
                        rDur = rState.dur
                    }
                    for (aState, aDur) in trgStates {
                        if aState == rState {
                            if aDur > State.STATE_END_DUR {
                                if aDur < 0 && rDur > aDur {
                                    return
                                }
                                rState.disable(actor: target, dur: rDur, remove: false)
                                if rDur > 0 && aDur > 0 {
                                    d -= aDur < rDur ? aDur : rDur
                                    if d < 1 {
                                        return
                                    }
                                }
                            }
                        }
                    }
                }
            }
            self.blockSkills(actor: target, remove: false)
            let crDur = trgStates[self] ?? State.STATE_END_DUR
            if crDur == State.STATE_END_DUR {
                self.adopt(actor: target, updStates: false, remove: false)
                trgStates[self] = d
            } else if (crDur > -1 && crDur < d) || (d < 0 && d < crDur) {
                trgStates[self] = d
            }
            if user != nil {
                if self.converted && target.side != user.side {
                    target.side = user.side
                }
                if self.covered {
                    target.coveredBy = user
                }
                if self.drawn {
                    target.drawnBy = user
                }
            }
        }
    }
    
    init(id: Int, name: String, sprite: String?, shapeShift: Bool, dur: Int, sRes: Int, mActions: Int, dmgType: Int,
         rflType: Int, hpDmg: Int, mpDmg: Int, spDmg: Int, mHp: Int, mMp: Int, mSp: Int, atk: Int, def: Int, spi: Int,
         wis: Int, agi: Int, stun: Bool, range: Bool, enrage: Bool, confuse: Bool, convert: Bool, reflect: Bool,
         invincible: Bool, cover: Bool, draw: Bool, ko: Bool, revive: Bool, aSkills: [Ability]?, rSkills: [Ability]?,
         counters: [Ability]?, states: [State: Int]?, stRes: [State: Int]?, res: [Int: Int]?) {
        self._dur = dur
        self._sRes = sRes
        self._rSkills = rSkills
        super.init(id: id, name: name, sprite: sprite, shapeShift: shapeShift, mActions: mActions, dmgType: dmgType, rflType: rflType,
                   hpDmg: hpDmg, mpDmg: mpDmg, spDmg: spDmg, mHp: mHp, mMp: mMp, mSp: mSp, atk: atk, def: def, spi: spi, wis: wis,
                   agi: agi, stun: stun, range: range, enrage: enrage, confuse: confuse, invincible: invincible, ko: ko, revive: revive,
                   skills: aSkills, counters: counters, states: states, stRes: stRes, res: res)
        var flags = self._flags
        if convert {
            flags |= State.FLAG_CONVERT
        }
        if cover {
            flags |= State.FLAG_COVER
        }
        if draw {
            flags |= State.FLAG_DRAW
        }
        self._flags = flags
    }
    
}
